File: typescript/core/test/humanGate.test.ts
/**
 * Unit tests for human-in-the-loop approval (HumanGate).
 *
 * Driven by the same fake OpenAI-compatible client the other agent tests use, so
 * no credentials are needed. Covers the three behaviors: an approved tool
 * executes; a denied tool does NOT execute and its denial reason reaches the
 * model; and with no gate configured behavior is unchanged.
 */

import { describe, expect, it } from 'vitest';
import { AgentOptions, ChatClientLike, SmoothAgent, Tool } from '../src/agent.js';
import { approve, deny, HumanApprovalRequest, HumanDecision } from '../src/humanGate.js';

type ScriptedMessage = {
    content: string | null;
    tool_calls?: Array<{ id: string; function: { name: string; arguments: string } }> | null;
};

class FakeClient implements ChatClientLike {
    readonly calls: Array<Record<string, unknown>> = [];
    private readonly scripted: ScriptedMessage[];

    constructor(scripted: ScriptedMessage[]) {
        this.scripted = [...scripted];
    }

    chat = {
        completions: {
            create: async (body: Record<string, unknown>) => {
                this.calls.push(body);
                const message = this.scripted.shift()!;
                return { choices: [{ message }] };
            },
        },
    };
}

/** A tool that records every invocation so a test can assert it never ran. */
function spyTool(): { tool: Tool; invocations: Array<Record<string, unknown>> } {
    const invocations: Array<Record<string, unknown>> = [];
    const tool: Tool = {
        name: 'delete_record',
        description: 'Deletes a record (destructive).',
        parameters: { type: 'object', properties: { id: { type: 'string' } }, required: ['id'] },
        async execute(args) {
            invocations.push(args);
            return `deleted record ${String(args.id ?? '')}`;
        },
    };
    return { tool, invocations };
}

function toolCall(id: string, name: string, args: string): ScriptedMessage {
    return { content: null, tool_calls: [{ id, function: { name, arguments: args } }] };
}

describe('HumanGate', () => {
    it('approve/deny helpers', () => {
        expect(approve().decision).toBe(HumanDecision.Approved);
        const d = deny('nope');
        expect(d.decision).toBe(HumanDecision.Denied);
        expect(d.reason).toBe('nope');
    });

    it('executes a tool when the gate approves', async () => {
        const { tool, invocations } = spyTool();
        const seen: HumanApprovalRequest[] = [];
        const client = new FakeClient([toolCall('c1', 'delete_record', '{"id":"42"}'), { content: 'done' }]);
        const options: AgentOptions = {
            tools: [tool],
            humanGate: async (req) => {
                seen.push(req);
                return approve();
            },
            requiresApproval: (name) => name === 'delete_record',
        };
        const agent = new SmoothAgent(client, options);
        const result = await agent.run('delete record 42');

        expect(result.text).toBe('done');
        expect(result.toolCalls).toBe(1);
        // The gate saw the right request, and the tool actually ran.
        expect(seen).toHaveLength(1);
        expect(seen[0].toolName).toBe('delete_record');
        expect(seen[0].arguments).toEqual({ id: '42' });
        expect(invocations).toEqual([{ id: '42' }]);
        // The successful tool result was fed back to the model.
        const secondCallMessages = client.calls[1].messages as Array<Record<string, unknown>>;
        expect(secondCallMessages.some((m) => m.role === 'tool' && String(m.content).includes('deleted record 42'))).toBe(true);
    });

    it('does not execute a denied tool and feeds the denial reason to the model', async () => {
        const { tool, invocations } = spyTool();
        const client = new FakeClient([toolCall('c1', 'delete_record', '{"id":"42"}'), { content: "understood, I won't delete it" }]);
        const options: AgentOptions = {
            tools: [tool],
            humanGate: async () => deny('policy forbids deletes'),
            requiresApproval: (name) => name === 'delete_record',
        };
        const agent = new SmoothAgent(client, options);
        const result = await agent.run('delete record 42');

        // The tool never ran.
        expect(invocations).toEqual([]);
        expect(result.text).toBe("understood, I won't delete it");
        // The denial (with reason) was fed back to the model as the tool result.
        const secondCallMessages = client.calls[1].messages as Array<Record<string, unknown>>;
        const denial = secondCallMessages.find((m) => m.role === 'tool');
        expect(denial).toBeDefined();
        expect(String(denial!.content)).toContain('Denied by human');
        expect(String(denial!.content)).toContain('policy forbids deletes');
    });

    it('leaves behavior unchanged when no gate is configured', async () => {
        const { tool, invocations } = spyTool();
        const client = new FakeClient([toolCall('c1', 'delete_record', '{"id":"42"}'), { content: 'done' }]);
        // No humanGate set — even though requiresApproval matches, it is ignored.
        const agent = new SmoothAgent(client, { tools: [tool], requiresApproval: () => true });
        const result = await agent.run('delete record 42');

        expect(result.text).toBe('done');
        expect(invocations).toEqual([{ id: '42' }]);
    });

    it('only consults the gate for flagged tools', async () => {
        const { tool, invocations } = spyTool();
        const consulted: string[] = [];
        const client = new FakeClient([toolCall('c1', 'delete_record', '{"id":"7"}'), { content: 'done' }]);
        const agent = new SmoothAgent(client, {
            tools: [tool],
            humanGate: async (req) => {
                consulted.push(req.toolName);
                return deny('should not be asked');
            },
            // Flags a different tool, so this one runs without consulting the gate.
            requiresApproval: (name) => name === 'send_email',
        });
        const result = await agent.run('delete record 7');

        expect(consulted).toEqual([]);
        expect(invocations).toEqual([{ id: '7' }]);
        expect(result.text).toBe('done');
    });
});

File: go/core/human_gate_test.go
package core

import (
	"context"
	"strings"
	"testing"
)

// gateSpyTool records every invocation so a test can assert it never ran.
func gateSpyTool(invocations *[]map[string]any) FuncTool {
	return FuncTool{
		ToolName: "delete_record",
		Desc:     "Deletes a record (destructive).",
		Params:   map[string]any{"type": "object", "properties": map[string]any{"id": map[string]any{"type": "string"}}, "required": []string{"id"}},
		Fn: func(_ context.Context, args map[string]any) (string, error) {
			*invocations = append(*invocations, args)
			id, _ := args["id"].(string)
			return "deleted record " + id, nil
		},
	}
}

func TestHumanResponseHelpers(t *testing.T) {
	if !Approve().IsApproved() {
		t.Fatal("Approve() should be approved")
	}
	d := Deny("nope")
	if d.IsApproved() || d.Decision != HumanDenied || d.Reason != "nope" {
		t.Fatalf("unexpected deny response: %+v", d)
	}
}

func TestApprovedToolExecutes(t *testing.T) {
	var invocations []map[string]any
	tool := gateSpyTool(&invocations)
	var seen []HumanApprovalRequest

	client := &fakeClient{scripted: []ChatResponse{
		{ToolCalls: []ToolCall{{ID: "c1", Name: "delete_record", Arguments: `{"id":"42"}`}}},
		{Content: "done"},
	}}
	agent := NewSmoothAgent(client, AgentOptions{
		Tools: []Tool{tool},
		HumanGate: func(_ context.Context, req HumanApprovalRequest) (HumanApprovalResponse, error) {
			seen = append(seen, req)
			return Approve(), nil
		},
		RequiresApproval: func(name string, _ map[string]any) bool { return name == "delete_record" },
	})

	res, err := agent.Run(context.Background(), "delete record 42", nil)
	if err != nil {
		t.Fatal(err)
	}
	if res.Text != "done" || res.ToolCalls != 1 {
		t.Fatalf("unexpected result: %+v", res)
	}
	// The gate saw the right request, and the tool actually ran.
	if len(seen) != 1 || seen[0].ToolName != "delete_record" {
		t.Fatalf("gate not consulted correctly: %+v", seen)
	}
	if id, _ := seen[0].Arguments["id"].(string); id != "42" {
		t.Fatalf("request args not passed through: %+v", seen[0].Arguments)
	}
	if len(invocations) != 1 {
		t.Fatalf("tool should have run once, ran %d times", len(invocations))
	}
	// The successful tool result was fed back to the model.
	second := client.calls[1].Messages
	found := false
	for _, m := range second {
		if m.Role == "tool" && strings.Contains(m.Content, "deleted record 42") {
			found = true
		}
	}
	if !found {
		t.Fatalf("tool result not fed back; messages=%+v", second)
	}
}

func TestDeniedToolDoesNotExecuteAndReasonReachesModel(t *testing.T) {
	var invocations []map[string]any
	tool := gateSpyTool(&invocations)

	client := &fakeClient{scripted: []ChatResponse{
		{ToolCalls: []ToolCall{{ID: "c1", Name: "delete_record", Arguments: `{"id":"42"}`}}},
		{Content: "understood, I won't delete it"},
	}}
	agent := NewSmoothAgent(client, AgentOptions{
		Tools: []Tool{tool},
		HumanGate: func(_ context.Context, _ HumanApprovalRequest) (HumanApprovalResponse, error) {
			return Deny("policy forbids deletes"), nil
		},
		RequiresApproval: func(name string, _ map[string]any) bool { return name == "delete_record" },
	})

	res, err := agent.Run(context.Background(), "delete record 42", nil)
	if err != nil {
		t.Fatal(err)
	}
	// The tool never ran.
	if len(invocations) != 0 {
		t.Fatalf("denied tool should not run; ran %d times", len(invocations))
	}
	if res.Text != "understood, I won't delete it" {
		t.Fatalf("unexpected final text: %q", res.Text)
	}
	// The denial (with reason) was fed back to the model as the tool result.
	second := client.calls[1].Messages
	var denial string
	for _, m := range second {
		if m.Role == "tool" {
			denial = m.Content
		}
	}
	if !strings.Contains(denial, "Denied by human") || !strings.Contains(denial, "policy forbids deletes") {
		t.Fatalf("denial reason not fed back; got %q", denial)
	}
}

func TestNoGateConfiguredLeavesBehaviorUnchanged(t *testing.T) {
	var invocations []map[string]any
	tool := gateSpyTool(&invocations)

	client := &fakeClient{scripted: []ChatResponse{
		{ToolCalls: []ToolCall{{ID: "c1", Name: "delete_record", Arguments: `{"id":"42"}`}}},
		{Content: "done"},
	}}
	// No HumanGate set — even though RequiresApproval matches, it is ignored.
	agent := NewSmoothAgent(client, AgentOptions{
		Tools:            []Tool{tool},
		RequiresApproval: func(string, map[string]any) bool { return true },
	})

	res, err := agent.Run(context.Background(), "delete record 42", nil)
	if err != nil {
		t.Fatal(err)
	}
	if res.Text != "done" || len(invocations) != 1 {
		t.Fatalf("expected unchanged behavior; result=%+v invocations=%d", res, len(invocations))
	}
}

func TestGateOnlyConsultsFlaggedTools(t *testing.T) {
	var invocations []map[string]any
	tool := gateSpyTool(&invocations)
	var consulted []string

	client := &fakeClient{scripted: []ChatResponse{
		{ToolCalls: []ToolCall{{ID: "c1", Name: "delete_record", Arguments: `{"id":"7"}`}}},
		{Content: "done"},
	}}
	agent := NewSmoothAgent(client, AgentOptions{
		Tools: []Tool{tool},
		HumanGate: func(_ context.Context, req HumanApprovalRequest) (HumanApprovalResponse, error) {
			consulted = append(consulted, req.ToolName)
			return Deny("should not be asked"), nil
		},
		// Flags a different tool, so this one runs without consulting the gate.
		RequiresApproval: func(name string, _ map[string]any) bool { return name == "send_email" },
	})

	res, err := agent.Run(context.Background(), "delete record 7", nil)
	if err != nil {
		t.Fatal(err)
	}
	if len(consulted) != 0 {
		t.Fatalf("gate should not be consulted for unflagged tool; consulted=%v", consulted)
	}
	if res.Text != "done" || len(invocations) != 1 {
		t.Fatalf("expected tool to run; result=%+v invocations=%d", res, len(invocations))
	}
}
